Clean up PersistentVolumeClaimSpec and PersistentVolumeSpec documentation

verult opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind documentation

#### What this PR does / why we need it:
Added more detailed explanations for certain fields, also cleaned up comments to improve consistency (following recommendations [here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#edit-typesgo))
* Removing mentions of the field name where possible
* Syncing comments across internal and external API definitions

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/sig storage
/sig docs
/assign @msau42

verult committed:
Clean up PersistentVolumeSpec documentation

verult committed:
Updating auto-generated files

verult commented:
/triage accepted
/priority important-longterm

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -425,25 +425,53 @@ type PersistentVolumeClaimList struct {
+	// If the value is "" (an empty string), the claim can only be bound to a
+	// PersistentVolume with no StorageClass (the storageClassName field is
+	// either unset or set to "").
+	// If this field is not set, the behavior depends on whether the
```
This might be a bit too much for the API reference. I think we can reduce this down to default storageclass if the admission controller is running vs not. But leave the details about how to set the default storageclass to the docs.

verult commented on pkg/apis/core/types.go:
```diff
@@ -425,25 +425,53 @@ type PersistentVolumeClaimList struct {
 	Selector *metav1.LabelSelector
-	// Resources represents the minimum resources required
+	// Represents the minimum resources the volume should have.
+	// More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#resources
```
Might be worth leaving both links in place in case more details are added in the future. WDYT?

verult commented on pkg/apis/core/types.go:
```diff
@@ -425,25 +425,53 @@ type PersistentVolumeClaimList struct {
 type PersistentVolumeClaimSpec struct {
-	// Contains the types of access modes required
+	// Contains the types of access modes the volume should have.
+	// More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#access-modes-1
```
Yeah I think so, updating

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -425,25 +425,51 @@ type PersistentVolumeClaimList struct {
+	// * ReadWriteOnce - can be mounted read/write mode to exactly 1 host.
+	// * ReadOnlyMany - can be mounted in read-only mode to many hosts.
+	// * ReadWriteMany - can be mounted in read/write mode to many hosts.
+	// More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#access-modes-1
```
Maybe note that access modes available will depend on what storage plugins are available in the cluster.

Do we want to clarify what it means for multiple access modes to be specified?

verult replied:
Or should we just say it's recommended to use 1 access mode only?

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -425,25 +425,51 @@ type PersistentVolumeClaimList struct {
+	// If this field is not set, the behavior depends on whether the
+	// DefaultStorageClass admission plugin is turned on.
+	// * If the admission plugin is turned on, the administrator may specify a
+	//   default StorageClass. This claim can be bound only to PersistentVolumes
```
What happens if no default storageclass exists?

verult replied:
Oops over-deleted things. I'm also planning to add back the behavior when multiple storageclasses exist

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -425,25 +425,51 @@ type PersistentVolumeClaimList struct {
 	// +optional
 	StorageClassName *string
-	// volumeMode defines what type of volume is required by the claim.
+	// Defines what type of volume is required by the claim.
```
It might be good to expand this one also to explain Filesystem vs Block

verult commented:
Addressed comments except explaining multiple access modes. I'm leaning towards just mentioning multiple access modes are not recommended on PVC. For PVs on the other hand, is it possible for a volume to have multiple access modes?

verult commented on pkg/apis/core/types.go:
```diff
@@ -330,8 +340,13 @@ type PersistentVolumeSpec struct {
-	// or to remain in raw block state. Value of Filesystem is implied when not included in spec.
+	// Describes the volume mode. Can be one of the following:
+	// * Filesystem - the volume contains a filesystem. If it doesn't,
+	//   Kubernetes creates a filesystem before mounting it for the first time.
```
I'll say "storage plugin"

verult commented on pkg/apis/core/types.go:
```diff
@@ -425,25 +425,51 @@ type PersistentVolumeClaimList struct {
-	// Contains the types of access modes required
+	// Contains the types of access modes the volume should have. Valid access
+	// modes include:
+	// * ReadWriteOnce - can be mounted read/write mode to exactly 1 host.
```
Too many places to copy paste 😅  thanks

verult commented:
Hopefully I got all the copy-paste spots this time. I explicitly left the access mode comments different between PVCSpec and PVSpec.

There's still an open question of what to say about having multiple access modes

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -305,21 +305,34 @@ type PersistentVolume struct {
 // PersistentVolumeSpec has most of the details required to define a persistent volume
 type PersistentVolumeSpec struct {
-	// Resources represents the actual resources of the volume
+	// A description of the persistent volume's resources and capacity.
```
I'm looking at the other capacity fields, it looks like the descriptions there are more along the lines of "Represents the available capacity of a ..."

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -305,21 +305,34 @@ type PersistentVolume struct {
+	// The actual volume backing the persistent volume.
 	PersistentVolumeSource
-	// AccessModes contains all ways the volume can be mounted
+	// Describes the volume's access modes. Possible values are:
```
What about something like "All the ways a volume can be accessed"

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -305,21 +305,34 @@ type PersistentVolume struct {
 	// +optional
 	ClaimRef *ObjectReference
-	// Optional: what happens to a persistent volume when released from its claim.
+	// What happens to a persistent volume when released from its claim.
```
PersistentVolume

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -328,13 +341,20 @@ type PersistentVolumeSpec struct {
+	//   first time.
+	// * Block - the volume will not be formatted with a filesystem and will
+	//   remain a raw block device.
+	// Value of Filesystem is the default when not included in spec.
```
How about "Defaults to Filesystem when unset".

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -328,13 +341,20 @@ type PersistentVolumeSpec struct {
 	MountOptions []string
-	// volumeMode defines if a volume is intended to be used with a formatted filesystem
-	// or to remain in raw block state. Value of Filesystem is implied when not included in spec.
+	// Describes the volume mode. Can be one of the following:
```
How about something like "Specifies how a volume will be accessed by a pod".

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -328,13 +341,20 @@ type PersistentVolumeSpec struct {
-	// volumeMode defines if a volume is intended to be used with a formatted filesystem
-	// or to remain in raw block state. Value of Filesystem is implied when not included in spec.
+	// Describes the volume mode. Can be one of the following:
+	// * Filesystem - the volume contains a filesystem. If it doesn't,
```
Maybe we should use the docs wording here. Something like:
```
The volume will be mounted into Pods into a directory. If the volume is backed by a block device and the device is empty, the storage driver creates a filesystem on the device before mounting it for the first time.
```

And for block:
```
The volume is presented to the pod as a block device, without any filesystem on it. The pod needs to understand how to use a raw block device.
```

verult committed:
Addressed comments

verult committed:
Generated files again

msau42 commented:
@verult lmk when this is ready for review

verult commented:
cc @humblec

verult committed:
Update internal API with field names in comment; fix line limits

verult commented:
Ready for review 👍

dims commented:
This PR is older than 4 weeks and needs as rebase.

Please rebase the PR against latest master and reopen if still needed.

/close
